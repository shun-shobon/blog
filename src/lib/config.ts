export type Config = {
  isProduction: boolean;
  articlesPath: string;
};

export function getConfig(): Config {
  const isProduction = process.env.NODE_ENV === "production";
  const articlesPath = isProduction ? "articles" : "test/articles";

  return {
    isProduction,
    articlesPath,
  };
}
