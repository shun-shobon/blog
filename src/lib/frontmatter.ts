import * as O from "fp-ts/Option";
import YAML from "yaml";

type Frontmatter = {
  frontmatter: unknown;
  content: string;
};

export function extractFrontmatter(raw: string): O.Option<Frontmatter> {
  const regexp = /^---\n(.+?)\n---\n(.+)/;
  const [, rawFrontmatter, content] = regexp.exec(raw) ?? [];

  if (rawFrontmatter == null || content == null) return O.none;

  const frontmatter: unknown = YAML.parse(rawFrontmatter);
  return O.some({ frontmatter, content });
}
