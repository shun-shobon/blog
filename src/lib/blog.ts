import * as path from "node:path";

import glob from "fast-glob";

import { getConfig } from "./config";

async function findArticles() {
  const { articlesPath } = getConfig();
  const articles = await glob(path.join(articlesPath, "*/main.md"));
}
