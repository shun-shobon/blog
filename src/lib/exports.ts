import * as fs from "node:fs/promises";
import * as path from "node:path";

import { Temporal } from "@js-temporal/polyfill";
import fg from "fast-glob";
import remarkFrontmatter from "remark-frontmatter";
import remarkGfm from "remark-gfm";
import remarkMath from "remark-math";
import remarkParse from "remark-parse";
import { unified } from "unified";

import { ARTICLE_DATABASE_NAME } from "@/config";

import type { ArticleDatabaseSerialized } from "./article";
import type { Article, ArticlePath } from "./plugins";
import {
  MARKDOWN_FILENAME,
  remarkArticle,
  remarkDescriptionList,
  remarkEmbed,
  remarkHeadingId,
  remarkRemovePosition,
  remarkResolveReference,
  remarkSection,
} from "./plugins";

export type ArticleDatabase = {
  articles: Article[];
};

export async function exportArticleDatabase(
  fromDir: string,
  imageExportDir: string,
  dataExportDir: string,
): Promise<void> {
  const articles = await processArticles(fromDir, imageExportDir);
  sortArticles(articles);

  const tags = getAllTags(articles);

  const databaseSerialized: ArticleDatabaseSerialized = {
    tags,
    articles,
  };

  const databasePath = path.join(dataExportDir, ARTICLE_DATABASE_NAME);

  await fs.mkdir(path.dirname(databasePath), { recursive: true });
  await fs.writeFile(databasePath, JSON.stringify(databaseSerialized));
}

async function processArticles(
  fromDir: string,
  toDir: string,
): Promise<Article[]> {
  const slugs = await findArticleSlugs(fromDir);
  const articlePaths = slugs.map((slug) => ({ slug, fromDir, toDir }));

  const articles = await Promise.all(
    articlePaths.map((articlePath) => processArticle(articlePath)),
  );

  return articles;
}

function getAllTags(articles: Article[]): string[] {
  const tagSet = new Set<string>();

  articles.forEach((article) => {
    article.tags.forEach((tag) => tagSet.add(tag));
  });

  return Array.from(tagSet);
}

function sortArticles(articles: Article[]): void {
  articles.sort((a, b) => {
    return Temporal.ZonedDateTime.compare(
      Temporal.ZonedDateTime.from(b.createdAt),
      Temporal.ZonedDateTime.from(a.createdAt),
    );
  });
}

async function findArticleSlugs(basePath: string): Promise<string[]> {
  const pattern = path.join(basePath, "*", MARKDOWN_FILENAME);
  const files = await fg(pattern);

  return files.map((filePath) => {
    return path.basename(path.dirname(filePath));
  });
}

async function processArticle(articlePath: ArticlePath): Promise<Article> {
  const processor = unified()
    .use(remarkParse)
    .use(remarkFrontmatter)
    .use(remarkGfm)
    .use(remarkMath)
    .use(remarkRemovePosition)
    .use(remarkResolveReference)
    .use(remarkEmbed)
    .use(remarkDescriptionList)
    .use(remarkHeadingId)
    .use(remarkSection)
    .use(remarkArticle, articlePath)
    .freeze();

  const filePath = path.join(
    articlePath.fromDir,
    articlePath.slug,
    MARKDOWN_FILENAME,
  );

  const fileContent = await fs.readFile(filePath, "utf-8");

  const article = await processor.run(processor.parse(fileContent));

  return article;
}
