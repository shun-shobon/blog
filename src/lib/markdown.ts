import * as fs from "node:fs/promises";
import * as path from "node:path";

import { Temporal } from "@js-temporal/polyfill";
import fg from "fast-glob";
import remarkFrontmatter from "remark-frontmatter";
import remarkGfm from "remark-gfm";
import remarkMath from "remark-math";
import remarkParse from "remark-parse";
import { unified } from "unified";

import { ARTICLE_SUMMARIES_FILE_NAME } from "@/config";

import type { Article, ArticlePath } from "./plugins";
import {
  MARKDOWN_FILENAME,
  remarkArticle,
  remarkDescriptionList,
  remarkEmbed,
  remarkHeadingId,
  remarkRemovePosition,
  remarkResolveReference,
  remarkSection,
} from "./plugins";

export type ArticleSummaries = {
  summaries: ArticleSummary[];
};

export type ArticleSummary = Pick<
  Article,
  | "title"
  | "plainTitle"
  | "slug"
  | "lead"
  | "createdAt"
  | "updatedAt"
  | "emoji"
  | "tags"
>;

export async function exportArticles(
  fromDir: string,
  imageExportDir: string,
  dataExportDir: string,
): Promise<void> {
  const articles = await processArticles(fromDir, imageExportDir);
  sortArticles(articles);

  await exportArticleSummaries(dataExportDir, articles);

  const promies = articles.map((article) =>
    exportArticle(dataExportDir, article),
  );
  await Promise.all(promies);
}

async function exportArticleSummaries(
  dataExportDir: string,
  articles: Article[],
): Promise<void> {
  const summaries = articles.map(createArticleSummary);
  const articleSummariesPath = path.join(
    dataExportDir,
    ARTICLE_SUMMARIES_FILE_NAME,
  );
  await fs.mkdir(dataExportDir, { recursive: true });
  await fs.writeFile(articleSummariesPath, JSON.stringify({ summaries }));
}

function createArticleSummary(article: Article): ArticleSummary {
  const { title, plainTitle, slug, lead, createdAt, updatedAt, emoji, tags } =
    article;
  return { title, plainTitle, slug, lead, createdAt, updatedAt, emoji, tags };
}

async function exportArticle(
  dataExportDir: string,
  article: Article,
): Promise<void> {
  const articlePath = path.join(dataExportDir, `${article.slug}.json`);
  await fs.writeFile(articlePath, JSON.stringify(article));
}

async function processArticles(
  fromDir: string,
  toDir: string,
): Promise<Article[]> {
  const slugs = await findArticleSlugs(fromDir);
  const articlePaths = slugs.map((slug) => ({ slug, fromDir, toDir }));

  const articles = await Promise.all(
    articlePaths.map((articlePath) => processArticle(articlePath)),
  );

  return articles;
}

function sortArticles(articles: Article[]): void {
  articles.sort((a, b) => {
    const compareCreatedAt = Temporal.PlainDate.compare(
      a.createdAt,
      b.createdAt,
    );

    if (compareCreatedAt !== 0) {
      return -compareCreatedAt;
    }

    if (a.updatedAt && b.updatedAt) {
      return -Temporal.PlainDate.compare(a.updatedAt, b.updatedAt);
    } else if (a.updatedAt && !b.updatedAt) {
      return -1;
    } else if (!a.updatedAt && b.updatedAt) {
      return 1;
    } else {
      return 0;
    }
  });
}

async function findArticleSlugs(basePath: string): Promise<string[]> {
  const pattern = path.join(basePath, "*", MARKDOWN_FILENAME);
  const files = await fg(pattern);

  return files.map((filePath) => {
    return path.basename(path.dirname(filePath));
  });
}

async function processArticle(articlePath: ArticlePath): Promise<Article> {
  const processor = unified()
    .use(remarkParse)
    .use(remarkFrontmatter)
    .use(remarkGfm)
    .use(remarkMath)
    .use(remarkRemovePosition)
    .use(remarkResolveReference)
    .use(remarkEmbed)
    .use(remarkDescriptionList)
    .use(remarkHeadingId)
    .use(remarkSection)
    .use(remarkArticle, articlePath)
    .freeze();

  const filePath = path.join(
    articlePath.fromDir,
    articlePath.slug,
    MARKDOWN_FILENAME,
  );

  const fileContent = await fs.readFile(filePath, "utf-8");

  const article = await processor.run(processor.parse(fileContent));

  return article;
}
