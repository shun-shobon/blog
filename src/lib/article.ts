import type { ArticleSummaries, ArticleSummary } from "@/lib/markdown";
import { ARTICLE_SUMMARIES_FILE_NAME } from "@/lib/markdown";

import type { Article } from "./plugins";

export async function getAllArticleSummaries(): Promise<ArticleSummary[]> {
  return (
    (await import(`data/${ARTICLE_SUMMARIES_FILE_NAME}`)) as ArticleSummaries
  ).summaries;
}

export async function getArticleSummariesByTag(
  tag: string,
): Promise<ArticleSummary[]> {
  const summaries = await getAllArticleSummaries();

  return summaries.filter((summary) => summary.tags.includes(tag));
}

export async function getAllTags(): Promise<string[]> {
  const summaries = await getAllArticleSummaries();

  const tags = new Set<string>();
  summaries.forEach((summary) => summary.tags.forEach((tag) => tags.add(tag)));

  return Array.from(tags);
}

export async function getArticle(slug: string): Promise<Article> {
  return (await import(`data/${slug}.json`)) as Article;
}
