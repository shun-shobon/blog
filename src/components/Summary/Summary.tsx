import { remark } from "remark";
import stripMarkdown from "strip-markdown";

import { type Article } from "../../lib/blog";

export type Props = {
  article: Article;
};

export const Summary = ({ article }: Props): JSX.Element => {
  const summary = remark()
    .use(stripMarkdown)
    .processSync(article.contents)
    .toString()
    .slice(0, 200)
    .concat("...");

  return (
    <article>
      <h2 className="text-xl text-cyan-500">
        <a
          href={`articles/${article.slug}`}
          className="cursor-pointer hover:underline decoration-current"
        >
          {article.title}
        </a>
      </h2>
      <p className="text-gray-500 flex flex-wrap gap-2">
        {article.tags.map((tag) => (
          <span key={tag}>#{tag}</span>
        ))}
      </p>
      <p>{summary}</p>
    </article>
  );
};
