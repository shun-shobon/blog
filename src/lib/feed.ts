import { Temporal } from "@js-temporal/polyfill";
import { Feed } from "feed";

import { ORIGIN, TITLE } from "@/config";
import { getAllArticleSummaries, getAllTags } from "@/lib/article";

import { createOgpImageUrl } from "./ogp-image";
import { getArticleUrl, getUrl } from "./utils";

export async function generateFeed(): Promise<Feed> {
  const summaries = await getAllArticleSummaries();
  const tags = await getAllTags();

  const lastUpdated = summaries[0]?.updatedAt ?? summaries[0]?.createdAt;
  const updated = lastUpdated
    ? new Date(
        Temporal.PlainDateTime.from(lastUpdated)
          .toZonedDateTime("Asia/Tokyo")
          .toInstant().epochMilliseconds,
      )
    : new Date(Temporal.Now.instant().epochMilliseconds);

  const feed = new Feed({
    title: TITLE,
    description:
      "しゅん🌙(@shun_shobon)のブログサイトです。技術記事からポエムまで色々書きます。",
    id: ORIGIN,
    link: ORIGIN,
    language: "ja-JP",
    favicon: getUrl("/favicon.ico").href,
    copyright: "Copyright © 2023 Shuntaro Nishizawa (しゅん🌙)",
    updated,
    author: {
      name: "Shuntaro Nishizawa (しゅん🌙)",
      link: "https://s2n.tech",
    },
  });

  summaries.forEach((summary) => {
    const date = new Date(
      Temporal.PlainDateTime.from(summary.updatedAt ?? summary.createdAt)
        .toZonedDateTime("Asia/Tokyo")
        .toInstant().epochMilliseconds,
    );
    const published = new Date(
      Temporal.PlainDateTime.from(summary.createdAt)
        .toZonedDateTime("Asia/Tokyo")
        .toInstant().epochMilliseconds,
    );

    feed.addItem({
      title: summary.plainTitle,
      id: getArticleUrl(summary.slug).href,
      link: getArticleUrl(summary.slug).href,
      image: {
        url: createOgpImageUrl(summary.plainTitle, summary.tags),
        type: "image/png",
      },
      description: summary.lead,
      category: summary.tags.map((tag) => ({ name: tag })),
      date,
      published,
    });
  });

  tags.forEach((tag) => {
    feed.addCategory(tag);
  });

  return feed;
}
