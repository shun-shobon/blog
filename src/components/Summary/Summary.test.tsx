/** @jest-environment jsdom */

import { render } from "@testing-library/react";

import { Summary } from "./Summary";
import { getArticle } from "../../lib/blog";
import * as E from "fp-ts/Either";

describe("Summary", () => {
  it("renders", async () => {
    const article = await getArticle(
      "tests/articles",
      "2021-12-05/test1/main.md",
    );

    expect(E.isRight(article)).toBeTruthy();
    if (E.isLeft(article)) return;

    const component = render(<Summary article={article.right} />);

    expect(component.getByText("Test Article 1")).toBeInTheDocument();

    expect(component.asFragment()).toMatchSnapshot();
  });
});
