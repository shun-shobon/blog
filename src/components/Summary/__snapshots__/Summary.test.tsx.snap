// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary renders 1`] = `
<DocumentFragment>
  <article
    class="p-4 rounded-lg shadow"
  >
    <h2
      class="text-xl text-cyan-500"
    >
      <a
        class="cursor-pointer hover:underline decoration-current"
        href="articles/2021-12-05-test1"
      >
        Test Article 1
      </a>
    </h2>
    <p
      class="text-gray-500 flex flex-wrap gap-2"
    >
      <span>
        #foo
      </span>
      <span>
        #bar
      </span>
    </p>
    <p>
      Test Article 1

Header 1

Header 2

Header 3

Header 4

Header 5

Header 6

List

item 1

item 2

item 3

Nested item 1

Nested item 2

Nested item 3

Ordered List

item 1

item 2

item 3

Bold Italic...
    </p>
  </article>
</DocumentFragment>
`;
